# 📘 Práctica de Python: Tipos de Datos Básicos

## Objetivos de Aprendizaje

Este notebook está diseñado para practicar la creación, manipulación y operación de los tipos de datos fundamentales en Python. El objetivo es construir una base sólida antes de pasar a estructuras de datos más complejas.

## Nivel Estimado

**Beginner → Intermediate**

Los ejercicios aumentan en dificultad, comenzando con operaciones básicas y terminando con manipulaciones más avanzadas y conceptos como la "truthiness".

## Contenidos que se Practicarán

* **Enteros (`int`):** Operaciones aritméticas, asignación.
* **Flotantes (`float`):** Precisión, operaciones de división.
* **Strings (`str`):** Creación, concatenación, f-strings, métodos (indexing, slicing, `.upper()`, `.lower()`, `.strip()`, `.replace()`).
* **Booleanos (`bool`):** Operadores lógicos (`and`, `or`, `not`), operadores de comparación (`==`, `!=`, `<`, `>`).
* **Type Casting:** Conversión explícita entre tipos (`int()`, `str()`, `float()`).

---

### 1. Creación de Variables Básicas

* **Nivel:** Básico
* **Objetivo:** Crear variables de los cuatro tipos de datos básicos.
* **Consigna:** Crea cuatro variables:
    1.  `mi_edad` (un entero) con el valor `25`.
    2.  `mi_altura` (un flotante) con el valor `1.78`.
    3.  `mi_nombre` (un string) con tu nombre.
    4.  `es_estudiante` (un booleano) con el valor `True`.
    
    Finalmente, imprime cada variable y su tipo usando `type()`.

* **Pista:** Recuerda que los strings usan comillas (`"` o `'`) y los booleanos usan mayúscula (`True` o `False`).

In [1]:
# Completa tu código aquí
mi_edad = 25
mi_altura = 1.78
mi_nombre = "Isaias Carrion"
es_estudiante = True

print(f"Edad: {mi_edad} (Tipo: {type(mi_edad)})")
print(f"Altura: {mi_altura} (Tipo: {type(mi_altura)})")
print(f"Nombre: {mi_nombre} (Tipo: {type(mi_nombre)})")
print(f"Es estudiante {es_estudiante} (Tipo: {type(es_estudiante)})")

Edad: 25 (Tipo: <class 'int'>)
Altura: 1.78 (Tipo: <class 'float'>)
Nombre: Isaias Carrion (Tipo: <class 'str'>)
Es estudiante True (Tipo: <class 'bool'>)


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
mi_edad = 25
mi_altura = 1.78
mi_nombre = "Alex"
es_estudiante = True

print(f"Edad: {mi_edad} (Tipo: {type(mi_edad)})")
print(f"Altura: {mi_altura} (Tipo: {type(mi_altura)})")
print(f"Nombre: {mi_nombre} (Tipo: {type(mi_nombre)})")
print(f"Es estudiante: {es_estudiante} (Tipo: {type(es_estudiante)})")

#### Explicación
* **Qué hace el código:** El código asigna valores a cuatro variables diferentes. `mi_edad` es un `int` (número entero). `mi_altura` es un `float` (número decimal). `mi_nombre` es un `str` (cadena de texto). `es_estudiante` es un `bool` (valor de verdad). La función `type()` se usa para verificar el tipo de dato de cada variable.
* **Errores comunes:** Olvidar las comillas en los strings (`mi_nombre = Alex`), lo que Python interpreta como otra variable. Escribir `true` o `false` en minúscula (en Python es `True` y `False`).

---

### 2. Operaciones Aritméticas (int)

* **Nivel:** Básico
* **Objetivo:** Practicar los operadores aritméticos básicos con enteros.
* **Consigna:** Dadas dos variables `a = 15` y `b = 4`, calcula e imprime lo siguiente:
    1.  La suma (`+`)
    2.  La resta (`-`)
    3.  La multiplicación (`*`)
    4.  La división (`/`) (Observa el tipo de dato resultante)
    5.  La división entera (`//`) (El cociente)
    6.  El módulo (`%`) (El resto de la división)
    7.  La potencia (`**`) (a elevado a la b)

In [2]:
# Completa tu código aquí
a = 15
b = 4

print(f"Suma (a + b): {a + b}")
print(f"Resta (a - b): {a - b}")
print(f"Multiplicación (a * b): {a * b}")
print(f"División (a / b): {a / b}")
print(f"División Entera (a // b): {a // b}")
print(f"Módulo (a % b): {a % b}")
print(f"Potencia (a ** b): {a ** b}")

Suma (a + b): 19
Resta (a - b): 11
Multiplicación (a * b): 60
División (a / b): 3.75
División Entera (a // b): 3
Módulo (a % b): 3
Potencia (a ** b): 50625


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
a = 15
b = 4

print(f"Suma (a + b): {a + b}")
print(f"Resta (a - b): {a - b}")
print(f"Multiplicación (a * b): {a * b}")
print(f"División (a / b): {a / b}")
print(f"División Entera (a // b): {a // b}")
print(f"Módulo (a % b): {a % b}")
print(f"Potencia (a ** b): {a ** b}")

#### Explicación
* **Qué hace el código:** Muestra los 7 operadores aritméticos básicos. 
* **Punto clave:** Nota que la división normal (`/`) **siempre** devuelve un `float` (ej: `15 / 4 = 3.75`), incluso si la división es exacta (`10 / 2 = 5.0`). La división entera (`//`) devuelve solo la parte entera y mantiene el tipo (`int` si ambos son `int`).
* **Errores comunes:** Confundir `/` y `//`. Usar `^` para la potencia (en Python `^` es el operador XOR; la potencia es `**`).

---

### 3. Concatenación de Strings

* **Nivel:** Básico
* **Objetivo:** Unir (concatenar) múltiples strings.
* **Consigna:** Tienes tres variables: `nombre = "Juan"`, `apellido = "Pérez"`, `ciudad = "Madrid"`. Crea una variable `presentacion` que contenga la frase "Hola, mi nombre es Juan Pérez y vivo en Madrid." usando estas variables.

* **Pista:** Usa el operador `+` para unir strings. No olvides añadir los espacios en blanco (`" "`) entre las variables.

In [3]:
# Completa tu código aquí
nombre = "Juan"
apellido = "Pérez"
ciudad = "Madrid"

presentacion = (
    "Hola, mi nombre es " + nombre + " " + apellido + " y vivo en " + ciudad + "."
)

print(presentacion)

Hola, mi nombre es Juan Pérez y vivo en Madrid.


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
nombre = "Juan"
apellido = "Pérez"
ciudad = "Madrid"

presentacion = (
    "Hola, mi nombre es " + nombre + " " + apellido + " y vivo en " + ciudad + "."
)
print(presentacion)

#### Explicación
* **Qué hace el código:** Utiliza el operador `+` para concatenar (pegar) varias cadenas de texto en una sola. Los espacios (`" "`) y los signos de puntuación (`"."`) deben ser añadidos explícitamente como strings.
* **Errores comunes:** Olvidar los espacios entre las variables, resultando en `"...es JuanPérez y..."`.
* **Variante más optimizada:** Usar f-strings, que veremos más adelante (ej: `f"Hola, mi nombre es {nombre} {apellido} y vivo en {ciudad}."`).

---

### 4. Input de Usuario y Type Casting

* **Nivel:** Básico
* **Objetivo:** Recibir datos del usuario y convertirlos al tipo correcto (`int`).
* **Consigna:** Pide al usuario que ingrese su año de nacimiento usando `input()`. Luego, calcula y muestra su edad aproximada. (Asume que el año actual es 2025).

* **Pista:** La función `input()` **siempre** devuelve un `str`. Necesitarás convertir ese string a un `int` usando `int()` antes de poder restar.

In [7]:
# Completa tu código aquí
año_actual = 2025
año_nacimiento_str = input("Ingresa tu año de nacimiento: ")
año_nacimiento = int(año_nacimiento_str)
edad_calculada = año_actual - año_nacimiento

print(f"Tienes o cumplirias {edad_calculada} años en {año_actual}")

Tienes o cumplirias 37 años en 2025


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
ANO_ACTUAL = 2025

# 1. Recibir el input (siempre es str)
ano_nacimiento_str = input("Ingresa tu año de nacimiento: ")

# 2. Convertir (cast) a int
ano_nacimiento_int = int(ano_nacimiento_str)

# 3. Calcular
edad_aproximada = ANO_ACTUAL - ano_nacimiento_int

print(f"Tienes aproximadamente {edad_aproximada} años.")

#### Explicación
* **Qué hace el código:** `input()` pausa el programa y espera que el usuario escriba algo. El valor recibido (`ano_nacimiento_str`) es un string (ej: `"1990"`). `int()` lo convierte en un número entero (`1990`), permitiendo realizar la resta.
* **Errores comunes:** Intentar restar directamente: `ANO_ACTUAL - ano_nacimiento_str`. Esto lanza un `TypeError`, ya que no se puede restar un `str` de un `int`.

---

### 5. Comparaciones Booleanas

* **Nivel:** Básico
* **Objetivo:** Usar operadores de comparación para obtener un resultado booleano.
* **Consigna:** Dadas `edad = 18` y `tiene_licencia = False`. Escribe expresiones booleanas para responder a estas preguntas (deben imprimir `True` o `False`):
    1.  ¿Es la persona mayor de edad (18 o más)?
    2.  ¿Tiene la persona exactamente 18 años?
    3.  ¿Tiene la persona una edad diferente a 25?
    4.  ¿Tiene la persona licencia?

* **Pista:** Usa `==` para "exactamente igual", `!=` para "diferente", y `>=` para "mayor o igual".

In [10]:
# Completa tu código aquí
edad = 18
tiene_licencia = False

es_mayor = edad >= 18
print(f"Es mayor de edad?: {es_mayor}")

tiene_18 = edad == 18
print(f"Tiene 18 años?: {tiene_18}")

es_diferente_25 = edad != 25
print(f"Edad es diferente a la edad de 25: {es_diferente_25}")

print(f"Tiene licencia?: {tiene_licencia}")

Es mayor de edad?: True
Tiene 18 años?: True
Edad es diferente a la edad de 25: True
Tiene licencia?: False


In [9]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
edad = 18
tiene_licencia = False

# 1. ¿Es la persona mayor de edad (18 o más)?
es_mayor = edad >= 18
print(f"¿Es mayor de edad?: {es_mayor}")

# 2. ¿Tiene la persona exactamente 18 años?
tiene_18 = edad == 18
print(f"¿Tiene 18 años?: {tiene_18}")

# 3. ¿Tiene la persona una edad diferente a 25?
es_diferente_25 = edad != 25
print(f"¿Edad diferente a 25?: {es_diferente_25}")

# 4. ¿Tiene la persona licencia?
print(f"¿Tiene licencia?: {tiene_licencia}")

¿Es mayor de edad?: True
¿Tiene 18 años?: True
¿Edad diferente a 25?: True
¿Tiene licencia?: False


#### Explicación
* **Qué hace el código:** Evalúa expresiones de comparación. Cada expresión (`edad >= 18`, `edad == 18`, `edad != 25`) se resuelve en un valor booleano (`True` o `False`).
* **Errores comunes:** Usar un solo igual (`=`) para comparar. `edad = 18` es *asignación* (guarda el valor). `edad == 18` es *comparación* (pregunta si es igual).

---

### 6. Repetición de Strings

* **Nivel:** Básico
* **Objetivo:** Usar el operador `*` para repetir un string.
* **Consigna:** Crea un "separador" visual. Define una variable `separador = "-"` y una variable `longitud = 50`. Crea una variable `linea_separadora` que contenga el `separador` repetido `longitud` veces. Imprímela.

* **Pista:** Puedes "multiplicar" un string por un entero.

In [11]:
# Completa tu código aquí
separador = "-"
longitud = 50

linea_separadora = separador * longitud

print(linea_separadora)

--------------------------------------------------


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
separador = "-"
longitud = 50

linea_separadora = separador * longitud
print(linea_separadora)

#### Explicación
* **Qué hace el código:** El operador `*` (multiplicación) cuando se usa con un `str` y un `int`, realiza una repetición. `"-" * 50` crea un nuevo string que contiene 50 guiones seguidos.
* **Errores comunes:** Intentar multiplicar dos strings (`"a" * "b"`) o un string por un float (`"a" * 2.5`). Ambos lanzarán un `TypeError`.

---

### 7. Casting de float a int

* **Nivel:** Básico
* **Objetivo:** Entender cómo se pierde información al convertir de `float` a `int`.
* **Consigna:** Tienes `precio_decimal = 19.99`. Convierte este precio a un entero usando `int()` y guárdalo en `precio_entero`. Imprime ambas variables y sus tipos.
* **Pregunta:** ¿Qué le pasó a la parte decimal?

In [12]:
# Completa tu código aquí
precio_decimal = 19.99
precio_entero = int(precio_decimal)

print(f"precio decimal: {precio_decimal} (Tipo: {type(precio_decimal)})")
print(f"precio entero: {precio_entero} (Tipo: {type(precio_entero)})")

precio decimal: 19.99 (Tipo: <class 'float'>)
precio entero: 19 (Tipo: <class 'int'>)


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
precio_decimal = 19.99
precio_entero = int(precio_decimal)

print(f"Precio decimal: {precio_decimal} (Tipo: {type(precio_decimal)})")
print(f"Precio entero: {precio_entero} (Tipo: {type(precio_entero)})")

#### Explicación
* **Qué hace el código:** `int(19.99)` convierte el flotante a entero.
* **Punto clave:** Al convertir de `float` a `int`, Python **trunca** el número, es decir, elimina toda la parte decimal. **No redondea** (19.99 no se convierte en 20, se convierte en 19). La parte decimal se pierde.

---

### 8. Casting de int a str (f-strings)

* **Nivel:** Básico / Intermedio
* **Objetivo:** Usar f-strings para formatear e insertar variables en un string (forma moderna de concatenar).
* **Consigna:** Usando las variables del Ejercicio 1 (`mi_edad = 25`, `mi_nombre = "Alex"`), crea la misma frase del Ejercicio 3, pero esta vez usando un **f-string**.
    La frase debe ser: "Hola, soy Alex y tengo 25 años."

* **Pista:** Un f-string comienza con una `f` antes de las comillas (ej: `f"Mi texto {variable}"`).

In [14]:
# Completa tu código aquí
mi_edad = 25
mi_nombre = "Alex"

frase = f"Hola, soy {mi_nombre} y tengo {mi_edad} años"
print(frase)

Hola, soy Alex y tengo 25 años


In [13]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
mi_edad = 25
mi_nombre = "Alex"

frase = f"Hola, soy {mi_nombre} y tengo {mi_edad} años."
print(frase)

Hola, soy Alex y tengo 25 años.


#### Explicación
* **Qué hace el código:** Los f-strings (strings literales formateados) son la forma preferida de insertar variables en strings. Python reemplaza automáticamente `{mi_nombre}` y `{mi_edad}` con sus valores.
* **Punto clave:** A diferencia de la concatenación con `+` (Ejercicio 3), los f-strings manejan el casting automáticamente. No necesitas hacer `str(mi_edad)` para unirlo al string; el f-string lo hace por ti.

---

### 9. Métodos de String: `.upper()`, `.lower()`, `.strip()`

* **Nivel:** Intermedio
* **Objetivo:** Limpiar y normalizar un string.
* **Consigna:** Tienes un string con formato inconsistente: `entrada_usuario = "  HoLa MuNdO  "`. 
    Realiza los siguientes pasos:
    1.  Crea `limpio` eliminando los espacios en blanco del principio y el final.
    2.  Crea `mayusculas` convirtiendo `limpio` todo a mayúsculas.
    3.  Crea `minusculas` convirtiendo `limpio` todo a minúsculas.
    Imprime las tres nuevas variables.

* **Pista:** Los métodos se llaman con un punto (ej: `variable.metodo()`).

In [16]:
# Completa tu código aquí
entrada_usuario = "  HoLa MuNdO  "

limpio = entrada_usuario.strip()
mayusculas = limpio.upper()
minusculas = limpio.lower()

print(f"Original: '{entrada_usuario}'")
print(f"Limpio: '{limpio}'")
print(f"Mayúsculas: '{mayusculas}'")
print(f"Minúsculas: '{minusculas}'")

Original: '  HoLa MuNdO  '
Limpio: 'HoLa MuNdO'
Mayúsculas: 'HOLA MUNDO'
Minúsculas: 'hola mundo'


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
entrada_usuario = "  HoLa MuNdO  "

# 1. Quitar espacios
limpio = entrada_usuario.strip()

# 2. Convertir a mayúsculas
mayusculas = limpio.upper()

# 3. Convertir a minúsculas
minusculas = limpio.lower()

print(f"Original: '{entrada_usuario}'")
print(f"Limpio: '{limpio}'")
print(f"Mayúsculas: '{mayusculas}'")
print(f"Minúsculas: '{minusculas}'")

#### Explicación
* **Qué hace el código:** `.strip()` elimina espacios, tabulaciones y saltos de línea del inicio y el final del string. `.upper()` devuelve una nueva copia del string en mayúsculas. `.lower()` devuelve una nueva copia en minúsculas.
* **Errores comunes:** Pensar que los métodos modifican el string original. Los strings en Python son **inmutables**. `entrada_usuario.strip()` *devuelve* un nuevo string, no modifica `entrada_usuario`. Por eso debemos asignar el resultado a una nueva variable.

---

### 10. Indexing de Strings (Acceder a caracteres)

* **Nivel:** Intermedio
* **Objetivo:** Obtener caracteres específicos de un string usando su índice.
* **Consigna:** Dada la palabra `palabra = "Python"`:
    1.  Imprime el primer carácter.
    2.  Imprime el último carácter.
    3.  Imprime el tercer carácter.

* **Pista:** El indexing en Python empieza en 0. El primer carácter es `[0]`. Para el último, puedes usar el índice `-1`.

In [17]:
# Completa tu código aquí
palabra = "Python"

primero = palabra[0]
print(f"Primer caracter: {primero}")
ultimo = palabra[-1]
print(f"Ultimo caracter: {ultimo}")
tercero = palabra[2]
print(f"Tercer caracter: {tercero}")

Primer caracter: P
Ultimo caracter: n
Tercer caracter: t


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
palabra = "Python"

# 1. Primer carácter (índice 0)
primero = palabra[0]
print(f"Primer carácter: {primero}")

# 2. Último carácter (índice -1)
ultimo = palabra[-1]
print(f"Último carácter: {ultimo}")

# 3. Tercer carácter (índice 2)
tercero = palabra[2]
print(f"Tercer carácter: {tercero}")

#### Explicación
* **Qué hace el código:** Usa corchetes `[]` para acceder a posiciones (índices) dentro del string.
    * `P y t h o n`
    * `0 1 2 3 4 5` (Índices positivos)
    * `-6-5-4-3-2-1` (Índices negativos)
* **Errores comunes:** `IndexError`. Ocurre si intentas acceder a un índice que no existe (ej: `palabra[6]` en "Python", cuyo último índice es 5).

---

### 11. Slicing de Strings (Rebanadas)

* **Nivel:** Intermedio
* **Objetivo:** Obtener sub-secciones (rebanadas) de un string.
* **Consigna:** Dado el string `url = "https://www.google.com"`:
    1.  Obtén el protocolo (los primeros 5 caracteres).
    2.  Obtén el dominio (desde el carácter 8 hasta el 17).
    3.  Obtén la extensión (los últimos 4 caracteres).

* **Pista:** La sintaxis es `[inicio:fin]`. `inicio` es inclusivo, `fin` es exclusivo. Si omites `inicio`, empieza desde 0. Si omites `fin`, va hasta el final.

In [18]:
# Completa tu código aquí
url = "https://www.google.com"

protocolo = url[0:5]
print(f"Protocolo: {protocolo}")

Protocolo: https


In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
url = "https://www.google.com"

# 1. Protocolo (índices 0, 1, 2, 3, 4)
protocolo = url[0:5]  # o [:5]
print(f"Protocolo: {protocolo}")

# 2. Dominio (índices 8 al 17)
dominio = url[8:18]  # 'google.com'
print(f"Dominio: {dominio}")

# 3. Extensión (últimos 4)
extension = url[-4:]
print(f"Extensión: {extension}")

#### Explicación
* **Qué hace el código:** El slicing `[inicio:fin]` extrae caracteres.
    * `[:5]` -> Desde el inicio (índice 0) hasta (sin incluir) el índice 5.
    * `[8:18]` -> Desde el índice 8 hasta (sin incluir) el índice 18.
    * `[-4:]` -> Desde el 4to carácter contando desde el final (índice -4) hasta el final del string.

---

### 12. Operadores Booleanos (`and`, `or`, `not`)

* **Nivel:** Intermedio
* **Objetivo:** Combinar múltiples condiciones booleanas.
* **Consigna:** Simula las condiciones para entrar a una atracción. 
    `edad = 14`
    `altura_cm = 160`
    `tiene_pase_vip = False`
    
    Evalúa e imprime las siguientes condiciones:
    1.  ¿Es mayor de 12 años **y** mide más de 150 cm?
    2.  ¿Es mayor de 18 años **o** mide más de 170 cm?
    3.  ¿Puede entrar gratis? (Solo si tiene pase VIP).
    4.  ¿No tiene pase VIP? (Usar `not`).

In [ ]:
# Completa tu código aquí
edad = 14
altura_cm = 160
tiene_pase_vip = False

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
edad = 14
altura_cm = 160
tiene_pase_vip = False

# 1. ¿Es mayor de 12 años Y mide más de 150 cm?
condicion_1 = (edad > 12) and (altura_cm > 150)
print(f"Condición 1 (and): {condicion_1}")

# 2. ¿Es mayor de 18 años O mide más de 170 cm?
condicion_2 = (edad > 18) or (altura_cm > 170)
print(f"Condición 2 (or): {condicion_2}")

# 3. ¿Puede entrar gratis? (Solo si tiene pase VIP)
condicion_3 = tiene_pase_vip
print(f"Condición 3 (pase): {condicion_3}")

# 4. ¿No tiene pase VIP? (Usar not)
condicion_4 = not tiene_pase_vip
print(f"Condición 4 (not): {condicion_4}")

#### Explicación
* **`and`:** Devuelve `True` solo si AMBAS condiciones son `True`. (En el ej. 1: `True and True` -> `True`).
* **`or`:** Devuelve `True` si AL MENOS UNA condición es `True`. (En el ej. 2: `False or False` -> `False`).
* **`not`:** Invierte el valor booleano. `not True` es `False`, y `not False` es `True`.

---

### 13. Conversión de Tipos Implícita (Coerción)

* **Nivel:** Intermedio
* **Objetivo:** Entender cómo Python maneja operaciones entre `int` y `float`.
* **Consigna:** Tienes `num_entero = 10` y `num_flotante = 5.5`. 
    1.  Calcula la suma `resultado = num_entero + num_flotante`.
    2.  Imprime el `resultado` y su `type()`.
* **Pregunta:** ¿De qué tipo es el resultado y por qué?

In [ ]:
# Completa tu código aquí
num_entero = 10
num_flotante = 5.5

In [ ]:
# --- SOLUCICIÓN (mostrar solo si el usuario lo pide) ---
num_entero = 10
num_flotante = 5.5

resultado = num_entero + num_flotante

print(f"Resultado: {resultado}")
print(f"Tipo del resultado: {type(resultado)}")

#### Explicación
* **Qué hace el código:** Suma un `int` y un `float`.
* **Punto clave (Coerción de tipo):** El resultado es `15.5` (un `float`). Cuando Python realiza una operación entre un `int` y un `float`, "promociona" (convierte implícitamente) al `int` a `float` (10 -> 10.0) para realizar la operación y evitar perder la información decimal. El resultado es siempre un `float`.

---

### 14. Operador `in` (Pertenencia en Strings)

* **Nivel:** Intermedio
* **Objetivo:** Verificar si un substring existe dentro de otro string.
* **Consigna:** Tienes `frase = "El rápido zorro marrón salta sobre el perro perezoso."`
    1.  Verifica si la palabra "zorro" está en la `frase`.
    2.  Verifica si la palabra "gato" está en la `frase`.
    3.  Verifica si la palabra "RÁPIDO" (en mayúsculas) está en la `frase`.

* **Pista:** El operador `in` devuelve `True` o `False`.

In [ ]:
# Completa tu código aquí
frase = "El rápido zorro marrón salta sobre el perro perezoso."

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
frase = "El rápido zorro marrón salta sobre el perro perezoso."

print(f"¿'zorro' está en la frase?: {"zorro" in frase}")
print(f"¿'gato' está en la frase?: {"gato" in frase}")
print(f"¿'RÁPIDO' está en la frase?: {"RÁPIDO" in frase}")

#### Explicación
* **Qué hace el código:** El operador `in` busca si la cadena de la izquierda existe exactamente dentro de la cadena de la derecha.
* **Punto clave:** La búsqueda es sensible a mayúsculas y minúsculas (Case-Sensitive). Por eso `"zorro" in frase` es `True`, pero `"RÁPIDO" in frase` es `False` (la frase contiene "rápido").

---

### 15. Método de String: `.replace()`

* **Nivel:** Intermedio
* **Objetivo:** Reemplazar partes de un string.
* **Consigna:** Tienes `saludo = "Hola, [NOMBRE], bienvenido a [CIUDAD]."`. 
    Reemplaza `"[NOMBRE]"` por `"Ana"` y `"[CIUDAD]"` por `"Barcelona"` para crear un saludo personalizado.

* **Pista:** `.replace("texto_viejo", "texto_nuevo")` devuelve un *nuevo* string. Puedes encadenar los reemplazos.

In [ ]:
# Completa tu código aquí
saludo_plantilla = "Hola, [NOMBRE], bienvenido a [CIUDAD]."

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
saludo_plantilla = "Hola, [NOMBRE], bienvenido a [CIUDAD]."

# Encadenamiento de métodos
saludo_personalizado = saludo_plantilla.replace("[NOMBRE]", "Ana").replace(
    "[CIUDAD]", "Barcelona"
)

print(f"Plantilla: {saludo_plantilla}")
print(f"Personalizado: {saludo_personalizado}")

#### Explicación
* **Qué hace el código:** `saludo_plantilla.replace("[NOMBRE]", "Ana")` crea un string temporal (`"Hola, Ana, bienvenido a [CIUDAD]."`). A ese string temporal, se le aplica el segundo `.replace("[CIUDAD]", "Barcelona")`.
* **Errores comunes:** (Nuevamente) Pensar que `.replace()` modifica la variable original. No lo hace, por eso debemos asignar el resultado final a `saludo_personalizado`.

---

### 16. Inmutabilidad de los Strings

* **Nivel:** Intermedio / Avanzado
* **Objetivo:** Demostrar que los strings son inmutables (no se pueden cambiar).
* **Consigna:** Tienes `mi_string = "Pato"`. Intenta cambiar el primer carácter (la 'P') por una 'G' usando indexing (`mi_string[0] = 'G'`).
    **Este ejercicio DEBE lanzar un error.** Lee el error y explica qué significa.

In [ ]:
# Completa tu código aquí
mi_string = "Pato"

# Intenta cambiar el primer carácter (esto fallará)
# mi_string[0] = 'G'

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
mi_string = "Pato"

try:
    mi_string[0] = "G"
except TypeError as e:
    print(f"ERROR: {e}")
    print("Explicación: Los strings son 'inmutables'.")

# La forma correcta de "modificar" es crear un NUEVO string:
nuevo_string = "G" + mi_string[1:]
print(f"String original: {mi_string}")
print(f"Nuevo string: {nuevo_string}")

#### Explicación
* **Qué hace el código:** El intento `mi_string[0] = 'G'` falla con un `TypeError: 'str' object does not support item assignment`.
* **Punto clave:** Esto significa que no puedes cambiar *partes* de un string una vez creado (son **inmutables**). Si quieres "cambiar" un string, debes crear uno nuevo (como `nuevo_string = 'G' + mi_string[1:]`, que concatena 'G' con el slicing 'ato').

---

### 17. Formateo Avanzado de f-strings (float)

* **Nivel:** Avanzado
* **Objetivo:** Usar f-strings para controlar el formato de números flotantes (precisión decimal).
* **Consigna:** Tienes una variable `pi = 3.14159265`. Imprime el valor de PI formateado de las siguientes maneras:
    1.  Con solo 2 decimales.
    2.  Con solo 4 decimales.
    3.  En un espacio total de 10 caracteres, alineado a la derecha, con 2 decimales.

* **Pista:** La sintaxis es `{variable:.Nf}` para N decimales. Para alineación: `{variable:[ancho].[precision]f}`.

In [ ]:
# Completa tu código aquí
pi = 3.14159265

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
pi = 3.14159265

# 1. Con 2 decimales
print(f"PI (2 decimales): {pi:.2f}")

# 2. Con 4 decimales
print(f"PI (4 decimales): {pi:.4f}")

# 3. Ancho 10, alineado derecha, 2 decimales
print(f"PI (formateado): {pi:>10.2f}")

#### Explicación
* **Qué hace el código:** Utiliza especificadores de formato dentro del f-string.
    * `{pi:.2f}`: Formatea la variable `pi` como un `f`loat con `.2` decimales (redondea a `3.14`).
    * `{pi:>10.2f}`: El `10` define un ancho total de 10 caracteres. El `>` alinea a la derecha (rellenando con espacios a la izquierda).

---

### 18. Slicing Avanzado (step)

* **Nivel:** Avanzado
* **Objetivo:** Usar el parámetro `step` (paso) en el slicing de strings.
* **Consigna:** Dada la frase `texto = "0a1b2c3d4e5f6g7h8i9j"`
    1.  Extrae solo los números (empezando en el índice 0, saltando de 2 en 2).
    2.  Extrae solo las letras (empezando en el índice 1, saltando de 2 en 2).
    3.  Invierte el string completo.

* **Pista:** La sintaxis es `[inicio:fin:paso]`. Un paso de `-1` invierte el string.

In [ ]:
# Completa tu código aquí
texto = "0a1b2c3d4e5f6g7h8i9j"

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
texto = "0a1b2c3d4e5f6g7h8i9j"

# 1. Solo números (inicio 0, paso 2)
numeros = texto[::2]
print(f"Números: {numeros}")

# 2. Solo letras (inicio 1, paso 2)
letras = texto[1::2]
print(f"Letras: {letras}")

# 3. Invertir el string (paso -1)
invertido = texto[::-1]
print(f"Invertido: {invertido}")

#### Explicación
* **Qué hace el código:**
    * `[::2]`: Omite `inicio` y `fin` (toma todo el string), pero usa un `paso` de 2. Toma los índices 0, 2, 4, ...
    * `[1::2]`: Empieza en el índice 1, toma todo el resto, con un `paso` de 2. Toma los índices 1, 3, 5, ...
    * `[::-1]`: Un `paso` negativo de -1 recorre el string al revés. Es el truco más común en Python para invertir un string.

---

### 19. Concepto de "Truthiness"

* **Nivel:** Avanzado
* **Objetivo:** Entender qué valores (no booleanos) evalúa Python como `True` o `False`.
* **Consigna:** Usa la función `bool()` para determinar el valor de verdad (Truthiness) de los siguientes elementos:
    1.  `valor_int_cero = 0`
    2.  `valor_int_positivo = 42`
    3.  `valor_float_cero = 0.0`
    4.  `valor_string_vacio = ""`
    5.  `valor_string_lleno = "Hola"`
    6.  `valor_none = None` (El tipo `NoneType`, que representa la ausencia de valor)

* **Pista:** Generalmente, los "ceros" o "vacíos" de cada tipo se evalúan como `False`.

In [ ]:
# Completa tu código aquí

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
valor_int_cero = 0
valor_int_positivo = 42
valor_float_cero = 0.0
valor_string_vacio = ""
valor_string_lleno = "Hola"
valor_none = None

print(f"bool(0): {bool(valor_int_cero)}")
print(f"bool(42): {bool(valor_int_positivo)}")
print(f"bool(0.0): {bool(valor_float_cero)}")
print(f"bool(\\"\\"): {bool(valor_string_vacio)}")
print(f"bool(\"Hola\"): {bool(valor_string_lleno)}")
print(f"bool(None): {bool(valor_none)}")

#### Explicación
* **Qué hace el código:** `bool()` muestra cómo Python interpreta diferentes valores en un contexto booleano (como en un `if`).
* **Punto clave (Valores "Falsy"):** En Python, los siguientes valores se consideran `False`:
    * `False`
    * `None`
    * `0` (int)
    * `0.0` (float)
    * `""` (string vacío)
    * `[]` (lista vacía), `{}` (diccionario vacío), `()` (tupla vacía)
* Cualquier otro valor (como `42` o `"Hola"`) se considera **"Truthy"** (evalúa como `True`).

---

### 20. Encadenamiento de Métodos (Method Chaining)

* **Nivel:** Avanzado
* **Objetivo:** Combinar múltiples métodos de string en una sola línea.
* **Consigna:** Tienes el siguiente string: `datos_sucios = "   ID_Usuario: 12345, Nombre: Juan   "`. 
    En una sola línea de código (asignada a `datos_limpios`), realiza las siguientes 3 acciones:
    1.  Elimina los espacios en blanco de los extremos.
    2.  Pon todo el texto en minúsculas.
    3.  Reemplaza el `"id_usuario: "` (nota el espacio) por `"id: "`.

    Imprime `datos_limpios`.

* **Pista:** `variable.metodo1().metodo2().metodo3()`

In [ ]:
# Completa tu código aquí
datos_sucios = "   ID_Usuario: 12345, Nombre: Juan   "

In [ ]:
# --- SOLUCIÓN (mostrar solo si el usuario lo pide) ---
datos_sucios = "   ID_Usuario: 12345, Nombre: Juan   "

# 1. .strip() -> 2. .lower() -> 3. .replace()
datos_limpios = datos_sucios.strip().lower().replace("id_usuario: ", "id: ")

print(f"Sucios: '{datos_sucios}'")
print(f"Limpios: '{datos_limpios}'")

#### Explicación
* **Qué hace el código:** Esto es posible porque cada método (`.strip()`, `.lower()`, `.replace()`) devuelve un *nuevo string*, sobre el cual se puede llamar inmediatamente al siguiente método.
* **Orden de ejecución:**
    1.  `datos_sucios.strip()` -> `"ID_Usuario: 12345, Nombre: Juan"`
    2.  *resultado_1*.`lower()` -> `"id_usuario: 12345, nombre: juan"`
    3.  *resultado_2*.`replace(...)` -> `"id: 12345, nombre: juan"`
* **Errores comunes:** El orden importa. Si hicieras `.lower()` antes que `.strip()`, funcionaría igual, pero si hicieras `.replace()` primero, fallaría, porque `replace` busca `"id_usuario: "` (minúsculas) y el string original aún estaría en mayúsculas (`"ID_Usuario: "`).